File: ts/src/day03.ts
import { DayBase } from './day-runner.js';

export class Day03 extends DayBase {
    constructor() {
        super(3);
    }

    public async run(): Promise<void> {
        const lines = await this.getInputLines();
        // const lines = ['00100','11110','10110','10111','10101','01111','00111','11100','10000','11001','00010','01010',]; // expect P1: 198 & P2: 230
        const blen = lines[0].length;

        // 'counts' contains the relative number of 1 (+ve) vs 0 (-ve) bits
        // in each position. ie if counts[2] > 0 then there are more 1s in
        // index position 2.
        let counts = Array<number>(blen).fill(0);
        lines.forEach(line => {
            for (let i = 0; i < blen; i++) {
                counts[i] += line[i] === '1' ? 1 : -1;
            }
        });
        // Treat +ve as 1 and -ve as 0 and convert to int
        let mc = 0, lc = 0;
        for (let i = 0; i < blen; i++) {
            if (counts[i] > 0) {
                mc += (1 << (blen- i - 1));
            } else {
                lc += (1 << (blen - i - 1));
            }
        }

        // Part 2
        let mfilt: string[] = [...lines], lfilt: string[] = [...lines]; //([] as string[]).concat(lines); //Array.from(lines);
        let idx = 0;
        while (mfilt.length > 1) {
            mfilt = this.filterValues(mfilt, idx++, true);
        }
        idx = 0;
        while (lfilt.length > 1) {
            lfilt = this.filterValues(lfilt, idx++, false);
        }
        const ogr = this.binaryStrToInt(mfilt[0]);
        const csr = this.binaryStrToInt(lfilt[0]);

        //console.log('mcb, mc, lc', mcb, mc, lc);
        //console.log('mfilt, lfilt', mfilt, lfilt);
        console.info(`Day ${this.day}`);
        console.info(`  Part 1:`, lc * mc);
        console.info(`  Part 2:`, ogr * csr);
        console.info('');
    }

    private binaryStrToInt(input: string): number {
        if (!input) { return 0; }
        const slen = input.length;
        let result = 0;
        for (let i = 0; i < slen; i++) {
            if (input[i] === '1') {
                result += (1 << (slen - i - 1));
            }
        }
        return result;
    }

    private filterValues(lines: string[], index: number, requireMostCommon: boolean): string[] {
        if (lines.length === 1) {
            return lines;
        }

        let ones: string[] = [];
        let zeros: string[] = [];
        lines.forEach(v => {
            if (v[index] === '1') {
                ones.push(v);
            } else {
                zeros.push(v);
            }
        })
        return requireMostCommon
            ? ones.length >= zeros.length ? ones : zeros   // most common
            : ones.length >= zeros.length ? zeros : ones;  // least common
    }
}

File: cpp/src/day03.cpp
#include <vector>
#include "aocutil.h"

using namespace std;

vector<int> countOnes(const vector<string> &src) {
    int bitwidth = src[0].size();
    vector<int> result(bitwidth, 0);
    for (auto &&num : src) {
        for (int i = 0; i < bitwidth; i++) {
            if (num[i] == '1') {
                result[i] += 1;
            }
        }
    }
    return result;
}

int countAt(const vector<string> &src, int bitPos) {
    int result = 0;
    for (auto &&num : src) {
        if (num[bitPos] == '1') {
            result++;
        }
    }
    return result;
}

vector<string> filterByBit(const vector<string> &src, int bitPos, bool requireLargest) {
    int ones = countAt(src, bitPos);
    int zeros = src.size() - ones;

    char seek = (requireLargest && ones >= zeros) || (!requireLargest && ones < zeros) ? '1' : '0';
    vector<string> result{};
    for (auto &&num : src) {
        if (num[bitPos] == seek) {
            result.push_back(num);
        }
    }
    return result;
}

string filterValues(const vector<string> &src, bool requireLargest) {
    int bp = 0;
    auto filt = src;
    while (filt.size() > 1) {
        filt = filterByBit(filt, bp, requireLargest);
        bp++;
    }
    return filt[0];
}

void day03()
{
    auto lines = get_input_lines(3, false);
    // Part 1
    auto counts = countOnes(lines);
    int half = lines.size() / 2;
    int gr = 0, er = 0;
    for (int i = 0; i < (int)counts.size(); i++) {
        gr *= 2;
        er *= 2;
        if (counts[i] >= half) {
            gr++;
        } else {
            er++;
        }
    }

    // Part 2
    int ogr = stoi(filterValues(lines, true).c_str(), nullptr, 2);
    int csr = stoi(filterValues(lines, false).c_str(), nullptr, 2);

	printDayResults(3, gr * er, ogr * csr);
}
